x-pack/legacy/plugins/index_management/static/ui/components/mappings_editor/mappings_state.tsx

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useReducer, useEffect, createContext, useContext } from 'react';
+
+import { reducer, MappingsConfiguration, MappingsFields, State, Dispatch } from './reducer';
+import { MAX_DEPTH_DEFAULT_EDITOR } from './constants';
+import { Field, FieldsEditor } from './types';
+import { normalize, deNormalize } from './lib';
+
+type Mappings = MappingsConfiguration & {
+  properties: MappingsFields;
+};
+
+export interface Types {
+  Mappings: Mappings;
+  MappingsConfiguration: MappingsConfiguration;
+  MappingsFields: MappingsFields;
+}
+
+export interface OnUpdateHandlerArg {
+  isValid?: boolean;
+  getData: () => Mappings;
+  validate: () => Promise<boolean>;
+}
+
+export type OnUpdateHandler = (arg: OnUpdateHandlerArg) => void;
+
+const StateContext = createContext<State | undefined>(undefined);
+const DispatchContext = createContext<Dispatch | undefined>(undefined);
+
+export interface Props {
+  children: (editor: FieldsEditor) => React.ReactNode;
+  defaultValue: { fields: { [key: string]: Field } };
+  onUpdate: OnUpdateHandler;
+}
+
+export const MappingsState = React.memo(({ children, onUpdate, defaultValue }: Props) => {
+  const { byId, rootLevelFields, maxNestedDepth } = normalize(defaultValue.fields);
```

Comment:
This is a very creative solution to the problem of traversing a nested data structure and then putting it back together once it's needed for mappings creation.

An alternate approach could be to have a mechanism for building the components tree and then closing over "parent", "children" and "selector" (i.e., the path to where a field is in the state tree) inside of the `DocumentFields` structure we are deriving from this. Then each "Add", "Remove" or "Edit" button would "know" where it is in the tree structure because of the order in which the tree structure is built. Not sure if that's clear - just wanted to share the thought.

Great work though! 😄

Reply:
Yeah, that was my first version... the problem started when you edited a field name anywhere down the tree. All the paths had to be updated. Unique ids removed all the headaches 😊